commit b83292413d860dec09ff14bf329e47a67ba30466
Author: Hongli Lai <hongli@phusion.nl>
Date:   Fri Nov 12 11:39:44 2010 +0100

    Apparently we can't always rely on 'struct dirent'->d_type for checking whether a directory entry is a directory, so fallback to normal stat() check in case it fails.

diff --git a/ext/common/ServerInstanceDir.h b/ext/common/ServerInstanceDir.h
index b21f68c..64a17a8 100644
--- a/ext/common/ServerInstanceDir.h
+++ b/ext/common/ServerInstanceDir.h
@@ -221,14 +221,18 @@ private:
 	}
 	
 	bool isDirectory(const string &dir, struct dirent *entry) const {
-		#ifdef __sun__
-			string path = dir;
-			path.append("/");
-			path.append(entry->d_name);
-			return getFileType(path) == FT_DIRECTORY;
-		#else
-			return entry->d_type == DT_DIR;
+		#ifdef DT_DIR
+			if (entry->d_type == DT_DIR) {
+				return true;
+			} else if (entry->d_type != DT_UNKNOWN) {
+				return false;
+			}
+			// If DT_UNKNOWN, use normal check.
 		#endif
+		string path = dir;
+		path.append("/");
+		path.append(entry->d_name);
+		return getFileType(path) == FT_DIRECTORY;
 	}
 	
 public:
